ddtrace/internal/module.py

```diff
@@ -0,0 +1,427 @@
+
+HookType = Callable[[ModuleType, Any], None]
+
+
+def origin(module):
+    # type: (ModuleType) -> str
+    """Get the origin of the module."""
+    try:
+        orig = abspath(module.__file__)  # type: ignore[type-var]
+    except (AttributeError, TypeError):
+        # Module is porbably only partially initialised, so we look at its
+        # spec instead
+        try:
+            orig = abspath(module.__spec__.origin)  # type: ignore
+        except (AttributeError, ValueError, TypeError):
+            orig = None
+
+    if orig is not None and isfile(orig):
+        if orig.endswith(".pyc"):
+            orig = orig[:-1]
+        return orig
+
+    return "<unknown origin>"
+
+
+def _resolve(path):
+    # type: (str) -> Optional[str]
+    """Resolve a (relative) path with respect to sys.path."""
+    for base in sys.path:
+        if isdir(base):
+            resolved_path = abspath(join(base, expanduser(path)))
+            if isfile(resolved_path):
+                return resolved_path
+    return None
+
+
+# Borrowed from the wrapt module
+# https://github.com/GrahamDumpleton/wrapt/blob/0baff1b6ac8d64ffb08ea2d1610c7ed90115b9d5/src/wrapt/importer.py
+class _ImportHookChainedLoader:
+    def __init__(self, loader, callback):
+        # type: (Loader, Callable[[ModuleType], None]) -> None
+        self.loader = loader
+        self.callback = callback
+
+    def load_module(self, fullname):
+        # type: (str) -> ModuleType
+        module = self.loader.load_module(fullname)
+        self.callback(module)
+
+        return module
+
+    def get_code(self, mod_name):
+        return self.loader.get_code(mod_name)
+
+
+class ModuleWatchdog(dict):
+    """Module watchdog.
+
+    Replace the standard ``sys.modules`` dictionary to detect when modules are
+    loaded/unloaded. This is also responsible for triggering any registered
+    import hooks.
+    """
+
+    _run_code = None
+
+    def __init__(self):
+        # type: () -> None
+        self._hook_map = defaultdict(list)  # type: DefaultDict[str, List[Tuple[HookType, Any]]]
+        self._origin_map = {origin(module): module for module in sys.modules.values()}
+        self._modules = sys.modules
+        self._finding = set()  # type: Set[str]
+
+    def __enter__(self):
+        # type: () -> ModuleWatchdog
+        self._add_to_meta_path()
+        sys.modules = self
+        log.debug("%s installed", type(self))
+        return self
+
+    def __exit__(self, *exc):
+        # type: (Tuple[Any]) -> None
+        self._remove_from_meta_path()
+        sys.modules = self._modules
+        log.debug("%s uninstalled", type(self))
+
+    def __getitem__(self, item):
+        # type: (str) -> ModuleType
+        return self._modules.__getitem__(item)
+
+    def __setitem__(self, name, module):
+        # type: (str, ModuleType) -> None
+        self._modules.__setitem__(name, module)
+
+    def _add_to_meta_path(self):
+        # type: () -> None
+        sys.meta_path.insert(0, self)
```

Comment:
I think this class would have to implement [`find_spec`](https://docs.python.org/3/library/importlib.html#importlib.abc.MetaPathFinder.find_spec) as well/instead of `find_module` which is deprecated as of Python 3.4 (I think we already have deprecation warnings emitted because our vendored version of wrapt only implements `find_spec`).

Also feels a tad off to have `ModuleWatchdog` implement the `MetaPathFinder` interface when also replacing the `sys.modules` interface. Something about being able to do `sys.modules.find_module` as a side-effect of using `ddtrace` irks me but not really sure why 😆

Reply:
I agree `ModuleWatchdog` is a `dict` _on steroids_. The main motivation for merging different interfaces was to reduce the number of classes needed to implement this feature. `find_module` is unfortunately required by Python 2.7. The fact that you can now call `sys.modules.find_module` is not super-nice, but to do that you would have to know about internal details of the library, which we don't support, so whoever does that will be doing it at their own risk.

Reply:
:+1: yeah I can't really come up with a solid argument against it other than some person discovering the method or depending on something crazy like `dir(sys.modules)`. So I think I'm ok with adding these methods.